doc/developer/design/20220411_reclocking_implementation.md

```diff
@@ -0,0 +1,303 @@
+
+* File
+* Kafka
+* Kinesis
+* PostgreSQL
+* S3
+
+The following sources types are nonreclockable:
+
+* PubNub
+
+### Storage layer implementation
+
+A reclockable source will consist of two definite collections in the storage
+layer: the [`remap`](20210714_reclocking.md#remap) collection, which contains
+the reclocking metadata, and the reclocked data collection.
+
+Nonreclockable sources will consist of only a single definite collection
+containing the source's data. For consistency with reclockable sources, we will
+refer to this collection as the "reclocked" collection, even though there is no
+upstream clock.
+
+The type of the `remap` collection depends on the source type:
+
+| Source type | `remap` data         | `remap` diff                  |
+|-------------|----------------------|-------------------------------|
+| File        | `ignored: ()`        | `byte_offset_increment: i64`  |
+| Kafka       | `partition: i32`     | `offset_increment: i64`       |
+| Kinesis     | `shard: i64`         | `sequence_increment: numeric` |
+| PostgreSQL  | `ignored: ()`        | `lsn_increment: i64`          |
+| S3          | `object_key: text`   | `byte_offset_increment: i64`  |
+
+The `remap` collection exists in the Materialize time domain, where a `u64`
+indicates milliseconds since the Unix epoch. In the future, if Materialize
+learns to support additional time domains, the `remap` collection will exist in
+the "target" time domain, which may be any one of the available time domains.
+
+Let's take a concrete example. Suppose the `remap` collection for a two
+partition Kafka source contains the following entries:
+
+```
+# (data,  time,           diff)
+  (1,     1649686076392,  +42)   # (1)
+  (2,     1649686076392,  +40)   # (2)
+  (1,     1649686079487,  +3)    # (3)
+  (2,     1649686079487,  +4)    # (4)
+```
+
+Entry (1) indicates that offsets 0-42 (inclusive) in partition 1 are assigned a
+Materialize timestamp of Monday, April 11, 2022 14:07:56.392 PM UTC. Entry (2)
+indicates that offsets 0-40 in partition 2 are assigned the same timestamp.
+Entry (3) indicates that offsets 43-45 in partition 1 are assigned a Materialize
+timestamp of 14:07:59.487 UTC on the same day. Entry (4) indicates that offsets
+41-44 in partition 2 are assigned the same timestamp.
+
+Notice how the offsets are described by their *increments*, not their absolute
+value. This allows the collection to be consolidated by summing the diffs of
+events with the same data.
+
+#### `remap` policy
+
+The remapping must be monotonic. Informally, the mapping must produce an order
+of events that does not conflict with order of events in the upstream source.
+See [`remap`](20210714_reclocking.md#remap) for a more formal statement of the
+property.
+
+We will impose an additional policy constraint on our implementation: if
+transactional boundaries are present in the upstream source, events in the same
+transaction must be mapped to the same timestamp.
```

Comment:
As it stands this property will either be true by definition or will not be achievable at reclock time. There are two cases:

1. The upstream transaction boundaries are encoded in the gauges themselves (like in pg lsn numbers) in which case any remap implementation upholds the property because each offset must be mapped to exactly one timestamp (modulo compaction)
2. The transaction boundaries are something more complex that we will learn about later in the pipeline (e.g decoding, envelope processing, or joining with some other reclocked stream in the case of debezium transaction topics). In this case I think the reclock operation can't possibly do the right thing and we shouldn't rely on it for the transaction boundaries. What happens instead (more or less) is that the ingestion stages later in the pipeline delay data accordingly and each transaction appears to happen when we receive the last part of it.

So I think I would remove this constraint from here

Reply:
> The upstream transaction boundaries are encoded in the gauges themselves (like in pg lsn numbers) in which case any remap implementation upholds the property because each offset must be mapped to exactly one timestamp (modulo compaction)

Doesn't each message have its own LSN? In the `wal_start` field of the `XLogDataBody` type? https://github.com/MaterializeInc/materialize/blob/07c8bd965b9928886d310fe36093cf13d08bfbb7/src/storage/src/source/postgres.rs#L404

We just (correclty) ignore it in our impl, and replace it with the end LSN of the transaction instead. This constraint is meant to assert that we keep doing that, and do the analogous thing for other sources we might add in the future.

> The transaction boundaries are something more complex that we will learn about later in the pipeline (e.g decoding, envelope processing, or joining with some other reclocked stream in the case of debezium transaction topics). In this case I think the reclock operation can't possibly do the right thing and we shouldn't rely on it for the transaction boundaries. What happens instead (more or less) is that the ingestion stages later in the pipeline delay data accordingly and each transaction appears to happen when we receive the last part of it.

We should talk about this with @frankmcsherry, because I think his vision is that the exposed remap collection would never contain a binding that splits a transaction in two, even if the downstream operators do the right thing. I don't think that's at odds with the new dbz txn stuff, since that is basically a remap operator that, IIUC, could occur _upstream_ of the timestamp bindings being produced. Are there other phases of decoding/de-enveloping that delay data?

Reply:
@benesch we actually don't do anything to do with mapping lsn's-timestamps right now, but "We just (correclty) ignore it in our impl, and replace it with the end LSN of the transaction instead." is what I'm going with in my `SourceReader for pg` change

Reply:
That sounds right, @guswynn, and that's all that this constraint is meant to enforce!

I filed #11952 so we can continue this discussion.